File: solutions/_77_combinations/solution.js
/**
 * @param {number} n
 * @param {number} k
 * @return {number[][]}
 */
var combine = function (n, k) {
  const res = [];
  backtrack(1, []);
  return res;

  function backtrack(start, path) {
    if (path.length === k) {
      res.push([...path]);
    } else {
      for (let i = start; i <= n; i++) {
        path.push(i);
        backtrack(i + 1, path);
        path.pop();
      }
    }
  }
};

module.exports = { combine };

/*
https://leetcode.com/problems/combinations/
*/

File: solutions/_77_combinations/solution.py
from typing import List


class Solution:
    def combine(self, n: int, k: int) -> List[List[int]]:
        def backtrack(start, path):
            if len(path) == k:
                res.append(path[:])
            else:
                for i in range(start, n + 1):
                    path.append(i)
                    backtrack(i + 1, path)
                    path.pop()

        res = []
        backtrack(1, [])
        return res


# https://leetcode.com/problems/combinations/
